pkg/controllers/clusterresourceplacement/clusterresourceplacement_controller.go

```diff
@@ -48,59 +81,119 @@ func (r *Reconciler) handleUpdate(ctx context.Context, crp *fleetv1beta1.Cluster
-				},
-			},
-			Spec: fleetv1beta1.SchedulingPolicySnapshotSpec{
-				Policy:     &schedulingPolicy,
-				PolicyHash: []byte(policyHash),
+
+	// create a new policy snapshot
+	latestPolicySnapshotIndex++
+	latestPolicySnapshot = &fleetv1beta1.ClusterSchedulingPolicySnapshot{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: fmt.Sprintf(fleetv1beta1.PolicySnapshotNameFmt, crp.Name, latestPolicySnapshotIndex),
+			Labels: map[string]string{
+				fleetv1beta1.CRPTrackingLabel:      crp.Name,
+				fleetv1beta1.IsLatestSnapshotLabel: strconv.FormatBool(true),
+				fleetv1beta1.PolicyIndexLabel:      strconv.Itoa(latestPolicySnapshotIndex),
 			},
+		},
+		Spec: fleetv1beta1.SchedulingPolicySnapshotSpec{
+			Policy:     &schedulingPolicy,
+			PolicyHash: []byte(policyHash),
+		},
+	}
+	policySnapshotKObj := klog.KObj(latestPolicySnapshot)
+	if err := controllerutil.SetControllerReference(crp, latestPolicySnapshot, r.Scheme); err != nil {
+		klog.ErrorS(err, "Failed to set owner reference", "clusterPolicySnapshot", policySnapshotKObj)
+		// should never happen
+		return nil, controller.NewUnexpectedBehaviorError(err)
+	}
+	// make sure each policySnapshot should always have the annotation if CRP is selectN type
+	if crp.Spec.Policy != nil &&
+		crp.Spec.Policy.PlacementType == fleetv1beta1.PickNPlacementType &&
+		crp.Spec.Policy.NumberOfClusters != nil {
+		latestPolicySnapshot.Annotations = map[string]string{
+			fleetv1beta1.NumberOfClustersAnnotation: strconv.Itoa(int(*crp.Spec.Policy.NumberOfClusters)),
 		}
-		policySnapshotKObj := klog.KObj(latestPolicySnapshot)
-		if err := controllerutil.SetControllerReference(crp, latestPolicySnapshot, r.Scheme); err != nil {
-			klog.ErrorS(err, "Failed to set owner reference", "clusterPolicySnapshot", policySnapshotKObj)
-			// should never happen
-			return ctrl.Result{}, controller.NewUnexpectedBehaviorError(err)
+	}
+
+	if err := r.Client.Create(ctx, latestPolicySnapshot); err != nil {
+		klog.ErrorS(err, "Failed to create new clusterPolicySnapshot", "clusterPolicySnapshot", policySnapshotKObj)
+		return nil, controller.NewAPIServerError(false, err)
+	}
+	return latestPolicySnapshot, nil
+}
+
+// TODO handle all the resources selected by placement larger than 1MB size limit of k8s objects.
+func (r *Reconciler) getOrCreateClusterResourceSnapshot(ctx context.Context, crp *fleetv1beta1.ClusterResourcePlacement, resourceSnapshotSpec *fleetv1beta1.ResourceSnapshotSpec) (*fleetv1beta1.ClusterResourceSnapshot, error) {
+	resourceHash, err := generateResourceHash(resourceSnapshotSpec)
+	if err != nil {
+		klog.ErrorS(err, "Failed to generate resource hash of crp", "clusterResourcePlacement", klog.KObj(crp))
+		return nil, controller.NewUnexpectedBehaviorError(err)
+	}
+
+	// latestResourceSnapshotIndex should be -1 when there is no snapshot.
+	latestResourceSnapshot, latestResourceSnapshotIndex, err := r.lookupLatestResourceSnapshot(ctx, crp)
```

Comment:
the resourceSnapshot and schedulingPolicySnapshot are generated following the same rule. Do we need to create two functions to look up the last? Will generic help unite them?

Reply:
the implementation will diverge when we handle the 1MB limit for the resourceSnapshot.

So prefer not to unify them, otherwise there will be too many "if else" conditions.